### PR TITLE
refactor[next]: Change foast lowering from iterator of tuple to tuple of iterator

### LINKED ISSUE #1413: tuples of fields of different locations violate ITIR
The following stencil from Icon4Py currently fails when using temporaries:
```python
@field_operator
def _mo_velocity_advection_stencil_17(
    e_bln_c_s: Field[[CEDim], wpfloat],
    z_v_grad_w: Field[[EdgeDim, KDim], vpfloat],
    ddt_w_adv: Field[[CellDim, KDim], vpfloat],
) -> Field[[CellDim, KDim], vpfloat]:
    z_v_grad_w_wp, ddt_w_adv_wp = z_v_grad_w, ddt_w_adv
    ...
```
The resulting ITIR contains the following:
```
(↑(λ(__arg0, __arg1) → {·__arg0, ·__arg1}))(z_v_grad_w, ddt_w_adv)
```
However since `z_v_grad_w` and `ddt_w_adv` live on different domains this is invalid. This currently breaks the temporary pass (as type inference fails). However when all lifts are force inlined the problem disappears so it never surfaced before.

### PR DESCRIPTION
**Pull request by tehrengruber**
Changes lowering from FOAST to ITIR to use tuples of iterators instead of iterators of tuples.
- Allows the collapse tuple pass to better optimize as what was previously lowered to `↑(λ(it) → it[0])(↑(λ() → {1, 2})` is now `({↑(λ() → 1)(), ↑(λ() → 2)()})[0]`
- Fixes #1413. In particular in combination with temporaries creating an iterator of tuples is problematic in case the respective backing fields are defined on different domains. This did not surface before as inlining all lifts everything always removed the invalid derefs.

### CODE REVIEW

File: src/gt4py/next/program_processors/runners/gtfn.py [hunk 1]
```diff
@@ -130,7 +130,7 @@ def compilation_hash(otf_closure: stages.ProgramCall) -> int:
 )
 
 GTFN_DEFAULT_COMPILE_STEP: step_types.CompilationStep = compiler.Compiler(
-    cache_strategy=cache.Strategy.SESSION, builder_factory=compiledb.CompiledbFactory()
+    cache_strategy=cache.Strategy.PERSISTENT, builder_factory=cmake.CMakeFactory()
```
Comment (tehrengruber) [inline src/gt4py/next/program_processors/runners/gtfn.py:133]: Will be removed before merge. Breaks code-quality so won't be forgotten.

File: src/gt4py/next/ffront/lowering_utils.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.type_system import type_info, type_specifications as ts
+
+
+def to_tuples_of_iterator(expr: itir.Expr | str, arg_type: ts.TypeSpec):
+    """
+    Convert iterator of tuples into tuples of iterator.
+
+    >>> _ = to_tuples_of_iterator("arg", ts.TupleType(types=[ts.FieldType(dims=[],
+    ...   dtype=ts.ScalarType(kind=ts.ScalarKind.FLOAT32))]))
```
Comment (havogt) [inline src/gt4py/next/ffront/lowering_utils.py:26]: the result might not be super stable, but would help to see the pretty printed from for understanding the code

File: src/gt4py/next/ffront/lowering_utils.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    )
+
+
+def to_iterator_of_tuples(expr: itir.Expr | str, arg_type: ts.TypeSpec):
+    """
+    Convert tuples of iterator into iterator of tuples.
+
+    >>> _ = to_iterator_of_tuples("arg", ts.TupleType(types=[ts.FieldType(dims=[],
+    ...   dtype=ts.ScalarType(kind=ts.ScalarKind.FLOAT32))]))
+    """
```
Comment (havogt) [inline src/gt4py/next/ffront/lowering_utils.py:50]: same here

File: src/gt4py/next/ffront/lowering_utils.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    >>> _ = to_tuples_of_iterator("arg", ts.TupleType(types=[ts.FieldType(dims=[],
+    ...   dtype=ts.ScalarType(kind=ts.ScalarKind.FLOAT32))]))
+    """
+    param = f"__toi_{abs(hash(expr))}"
+
+    def fun(primitive_type, path):
+        inner_expr = im.deref("it")
+        for path_part in path:
+            inner_expr = im.tuple_get(path_part, inner_expr)
```
Comment (havogt) [inline src/gt4py/next/ffront/lowering_utils.py:33]: does this deal with nesting? maybe extend the docstring

Comment (havogt) [reply src/gt4py/next/ffront/lowering_utils.py:33]: if that is nesting, we should add a few tests for this function

File: src/gt4py/next/ffront/lowering_utils.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        return im.lift(im.lambda_("it")(inner_expr))(param)
+
+    return im.let(param, expr)(
+        type_info.apply_to_primitive_constituents(
+            arg_type, fun, with_path_arg=True, tuple_constructor=im.make_tuple
+        )
+    )
+
+
+def to_iterator_of_tuples(expr: itir.Expr | str, arg_type: ts.TypeSpec):
```
Comment (havogt) [inline src/gt4py/next/ffront/lowering_utils.py:44]: and also here some tests

File: src/gt4py/next/ffront/past_to_itir.py [hunk 6]
```diff
@@ -141,16 +142,42 @@ def _visit_stencil_call(self, node: past.Call, **kwargs) -> itir.StencilClosure:
```
Comment (havogt) [inline src/gt4py/next/ffront/past_to_itir.py:142]: not really part of this pr, but maybe let's change the function signature to
```
def _visit_stencil_call(self, node: past.Call, *, out, domain=None, **kwargs) -> itir.StencilClosure:
```
to avoid copying kwargs, to avoid accidentally using `node.kwargs` instead of `node_kwargs`

File: src/gt4py/next/ffront/past_to_itir.py [hunk 7]
```diff
@@ -141,16 +142,42 @@ def _visit_stencil_call(self, node: past.Call, **kwargs) -> itir.StencilClosure:
...
+            # scan operators return an iterator of tuples
+            stencil_body = im.deref(im.call(node.func.id)(*stencil_args))
+        else:
+            # field operators return a tuple of iterators
+            # TODO(tehrengruber): change into deref per element
+            stencil_body = im.deref(
+                lowering_utils.to_iterator_of_tuples(
+                    im.call(node.func.id)(*stencil_args), node.func.type.definition.returns
+                )
+            )
```
Comment (havogt) [inline src/gt4py/next/ffront/past_to_itir.py:176]: why not implement the todo? seems like all utils are there?

File: src/gt4py/next/iterator/ir_utils/ir_makers.py [hunk 8]
```diff
@@ -250,10 +250,10 @@ class let:
...
 
     @typing.overload
-    def __init__(self, var: str | itir.Sym, init_form: itir.Expr): ...
+    def __init__(self, var: str | itir.Sym, init_form: itir.Expr | str): ...
 
     @typing.overload
-    def __init__(self, *args: Iterable[tuple[str | itir.Sym, itir.Expr]]): ...
+    def __init__(self, *args: Iterable[tuple[str | itir.Sym, itir.Expr | str]]): ...
 
     def __init__(self, *args):
```
Comment (havogt) [inline src/gt4py/next/iterator/ir_utils/ir_makers.py:258]: does type checking work properly if the implementation is not properly annotated with the union of the above?

**Review by havogt — CHANGES_REQUESTED** [review]
Can you carefully check where tests are missing? Difficult to follow the transformations without examples.

File: src/gt4py/next/ffront/foast_to_itir.py [hunk 9]
```diff
@@ -38,6 +39,55 @@ def promote_to_list(
     return lambda x: x
 
 
+# TODO(tehrengruber): The interface and code quality of this function is poor. We should rewrite it.
```
[the remark this answers is not in the document]
Comment (tehrengruber) [reply src/gt4py/next/ffront/foast_to_itir.py:42]: Very true. Done

File: src/gt4py/next/ffront/lowering_utils.py (hunk 4 above)
Comment (tehrengruber) [reply src/gt4py/next/ffront/lowering_utils.py:33]: Yes, can I skip the tests? We have integration tests covering this and the function is superfluous as soon as we go to combined IR.

File: src/gt4py/next/ffront/past_to_itir.py (hunk 6 above)
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_to_itir.py:142]: Read again, `out` and `domain` are not coming from `kwargs`, but are part of the `node.kwargs`. (Or I misunderstand you?)

File: src/gt4py/next/iterator/ir_utils/ir_makers.py (hunk 8 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/ir_utils/ir_makers.py:258]: Likely no one ever used `let(sym_a, sym_b)(...)` with `sym_b` being a string (as such rebinding cases usually mean `sym_b` is already an `itir.Sym` coming from somewhere). I took your suggestion.

File: src/gt4py/next/ffront/past_to_itir.py (hunk 7 above)
Comment (tehrengruber) [reply src/gt4py/next/ffront/past_to_itir.py:176]: True I decided to keep effort low (and honestly also motivation as I am in this continuous mini-change, retest cycle) as these functions will become superfluous with combined IR. If you would like to have it I'll do it.

**Comment by tehrengruber** [comment]
Let's quickly talk about testing tomorrow.

File: src/gt4py/next/iterator/ir_utils/ir_makers.py (hunk 8 above)
Comment (havogt) [reply src/gt4py/next/iterator/ir_utils/ir_makers.py:258]: You mean not annotating?

**Review by havogt — APPROVED** [review]
